Topicモデルを作成

tatsukishiga opened the pull request:
モデルを作成し、バリデーション、アソシエーションを追加した。

tatsukishiga committed:
Topicモデルを作成

モデルを作成し、バリデーション、アソシエーションを追加した。

tatsukishiga committed:
トピック一覧機能を作成

コントローラーを作成し、トピックの一覧ページを作成した。リンクも追加した。

tatsukishiga committed:
トピック作成機能を追加

作成機能を追加し、リンクも追加した。

tatsukishiga committed:
トピック作成のリンクを追加

リンクが追加出来てなかったので追加した。

ArakawaShintaro commented on app/controllers/topics_controller.rb:
```diff
@@ -0,0 +1,23 @@
+class TopicsController < ApplicationController
+  def index
+    @topics = Topic.all.order(created_at: :desc)
```
`Topic.order(created_at: :desc)`

でいけませんかね？

tatsukishiga replied:
試してみたら、そちらで出来ました。order単体で選択したテーブルの情報を全て取れると理解できてませんでした。修正します！

ArakawaShintaro commented on app/controllers/topics_controller.rb:
```diff
@@ -0,0 +1,23 @@
+  end
+
+  def create
+    @topic = Topic.new(
```
`@current_user.topic.new(~~~)`
とすると、user_idがセットされた状態になりますよー

tatsukishiga replied:
新しいトピックの作成をしているのは、ログインしているユーザーだから、これでいいのではないのでしょうか？

ArakawaShintaro commented on test/controllers/topics_controller_test.rb:
```diff
@@ -0,0 +1,9 @@
+require 'test_helper'
```
テスト系のファイルは消してください

tatsukishiga replied:
消します！

tatsukishiga committed:
トピックの詳細ページを追加

トピックの詳細ページを追加しました。

tatsukishiga committed:
トピックの編集ページを追加

編集ページの追加をしました。

tatsukishiga committed:
トピックの検索機能を追加

form_tagとwhereを用いてトピックの検索機能を書いてみましたがsearch_result画面の何を検索したかの表示の部分がかなり自信ないです。whereで取得出来なかった場合にも何を検索したのか表示できていないです；；

tatsukishiga committed:
ページネーションの追加

トピック一覧ページにページネーションを追加した。1ページあたり5項目表示されるようにした。

tatsukishiga committed:
検索結果ページ修正

検索結果ページにif文で検索件数が０の場合とそうでない場合によって表示を変えた。

tatsukishiga committed:
トピックのviewのHTMLを修正

HTMLの構成を修正をした。またトピック詳細ページにはトピックへのコメントの日時の表示を追加したのと、投稿したユーザーが削除できる機能を追加。

tatsukishiga committed:
テストファイルの削除

残りのテストファイルを削除しました。

tatsukishiga committed:
指摘箇所修正

controllerのアクションの書き方を修正。modelの不必要なコメント削除。viewのリンクのif文を修正。

tatsukishiga committed:
Resourcesでルーティングを定義

Resourcesメソッドを使ってルーティングを定義し直した。そのため、リンクやformタグ、リダイレクトで飛ばす先のコードを修正した。また不要なものを削除、インデントの修正をした。

ArakawaShintaro commented on app/controllers/topics_controller.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  def new
+    @topic = Topic.new
```
これ、口頭でも言いましたが
URLを直打ちされたらnewページへ言ってしまうと思います

tatsukishiga replied:
beforeアクションで対応しました。

ArakawaShintaro commented on app/controllers/topics_controller.rb:
```diff
@@ -0,0 +1,36 @@
+
+  def show
+    @topic = Topic.find_by(id: params[:id])
+    @user = @topic.user
```
@topic_userとかの方が明示できていいかもですね

tatsukishiga replied:
@topic_userに修正しました。

ArakawaShintaro commented on app/views/layouts/application.html.erb:
```diff
@@ -10,6 +10,20 @@
+    <header>
+      <div class="header-logo">
+        <% if @current_user %>
+          <%= link_to("Board", "/topics") %>
```
この際なので `/topics` ではなく名前付きpathでやってみましょう

tatsukishiga replied:
名前付きpathに変更しました

ArakawaShintaro commented on app/views/topics/edit.html.erb:
```diff
@@ -0,0 +1,21 @@
+<div class="main topic-new">
+  <div class="container">
+    <h1 class="form-heading">トピック編集</h1>
+    <%= form_tag("/topics/#{@topic.id}", {method: "patch"}) do %>
```
form_tagではなく、form_withでやりましょう

tatsukishiga replied:
form_withに修正しました。

tatsukishiga committed:
指摘箇所修正

1.URLで直打された場合の対策として、ログインを必要とするbeforeアクションを追加。
2.飛ばすURLにpathを使用した
3.showアクションの@topicを@topic_userに修正した。
4.editビューのform_tagをform_withに変更
5.ビュー内でのちょっとした処理をhelperを使用してメソッド化した。
6.updateアクションを追加
7.一旦ロールバックしてからdb側にバリデーションを追加した

tatsukishiga committed:
指摘箇所修正

1.form_tagをform_withに変更。
2.飛ばすURLにpathを使用。
3.ビュー内での処理をhelperを使用してメソッド化。メソッドに引数を渡す。

tatsukishiga committed:
追加修正

名前付きpath一部抜けていたため修正の追加

tatsukishiga committed:
修正追加

form_withで送られてきたものが受け取れていなかったのでparamsを追加して修正。

ArakawaShintaro commented on app/controllers/topics_controller.rb:
```diff
@@ -0,0 +1,49 @@
+  def update
+    @topic = Topic.find_by(id: params[:id])
+    @topic.update(title: params[:topic][:title], content: params[:topic][:content])
+    if @topic.update(title: params[:topic][:title], content: params[:topic][:content])
```
L37でも同じことをしている気が

ArakawaShintaro commented on app/controllers/topics_controller.rb:
```diff
@@ -0,0 +1,49 @@
+    @topic.update(title: params[:topic][:title], content: params[:topic][:content])
+    if @topic.update(title: params[:topic][:title], content: params[:topic][:content])
+      flash[:notice] = "トピックの情報を編集しました"
+      redirect_to("/topics/#{@topic.id}")
```
名前付きパス

ArakawaShintaro commented on app/helpers/topics_helper.rb:
```diff
@@ -0,0 +1,9 @@
+
+  def comment_user(user_id: comment.user_id)
+    user_id == @current_user.id
+  end
```
```ruby
 def comment_user?(comment: comment)
    comment.user_id == @current_user.id
  end
```

の方が自然な気がします
あとtrue/falseを返すメソッドは？をつけてあげるのが慣習です

ArakawaShintaro commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  # topics
+  resources :topics
+  post 'topics/search_result' => 'topics#search_result'
```
postでやった理由ってありますか？

tatsukishiga committed:
指摘事項修正

1.名前付きpathに出来るところを全て直した。renderについてはこの記述が一番わかりやすくシンプルかなと思いましたのでこちらで定義させていただきました。
2.helper周りの修正
3.update周りの重複を削除
4.コメントの編集リンクを忘れていたので追加

tatsukishiga committed:
検索機能のHTTPメソッドを変更

１，メソッドを使用し、post→getに変更２，前回の記述だとルーティングでshowが優先されて処理されエラーになるため、上にコードを移した。

ArakawaShintaro reviewed (approved):
LGTM

tatsukishiga committed:
Merge branch 'master' into topic_function
